File: tests/fixtures/timeseries/verify_arima.R
# verify_arima.R - ARIMA 基准生成: R stats::arima CSS/CSS-ML (Phase 7C M1)
#
# 对照点 (spec §1.3, AR6 配对):
#   ARIMA CSS vs R stats::arima method="CSS"    1e-10 (参数级实测校准)
#   ARIMA CSS-ML vs R method="CSS-ML"           1e-8
#   d=1 drift vs forecast::Arima (AR5)          1e-8
#
# 语义对齐 (一手实测留痕):
#   - d=0 夹具: include.mean=FALSE (C++ CSS 无均值项, spec 决策 3;
#     数据本身零均值生成 + demean 语义留给 innovations 对照)
#   - AR2 复核点: R CSS 的 n.cond 实测 = max(p, q)+? — 本脚本对 CSS 残差
#     长度/数值打印, 若与 C++ (n_cond−d = p 起) 不一致 → 以 R 实测为准回写
#   - MA 正号 (AR1): R coef 里 ma1 即 (1+θB) 的 θ
#   - loglik: R 纯 CSS 报 "part log-likelihood" (AR3 不与 ML 混比, 只对
#     CSS-ML/ML 的 loglik 数值对照)
#
# 用法: Rscript verify_arima.R  (读 arima_smoke_data.csv, 输出 12 位基准)
.libPaths(c('F:/R/win-library/4.6', file.path(Sys.getenv("USERPROFILE"), "R", "win-library", "4.6"), .libPaths()))

here <- commandArgs(trailingOnly = FALSE)
here <- dirname(normalizePath(sub("^--file=", "", here[grepl("^--file=", here)])))
csv <- file.path(here, "arima_smoke_data.csv")
# keys 从原始行读 (read.csv 会把 #columns 行按逗号拆列, 2026-08-18 实测)
lns <- readLines(csv)
keys <- strsplit(sub("^#columns=", "", lns[grepl("^#columns", lns)]), ",")[[1]]
raw <- read.csv(csv, header = FALSE, stringsAsFactors = FALSE)
dat <- raw[!grepl("^#", raw[[1]]), ]
cat("columns:", paste(keys, collapse = ", "), "\n")
arma11 <- as.numeric(dat[[1]])
arma22 <- as.numeric(dat[[2]])
arma21 <- as.numeric(dat[[3]])
arma12 <- as.numeric(dat[[4]])
arima111d_level <- as.numeric(dat[[5]])
stopifnot(!any(is.na(arima111d_level)))

fmt <- function(x) format(x, digits = 15, scientific = TRUE)

cat("=== arma11 (T=250) stats::arima CSS / CSS-ML, include.mean=FALSE ===\n")
a <- list()
for (m in c("CSS", "CSS-ML")) {
  fit <- arima(arma11, order = c(1, 0, 1), method = m, include.mean = FALSE)
  a[[m]] <- fit
  cat(m, ":")
  cat(" coef=", paste(fmt(fit$coef), collapse = ", "))
  cat(" sigma2=", fmt(fit$sigma2))
  cat(" loglik=", fmt(as.numeric(logLik(fit))))
  cat(" aic=", fmt(fit$aic))
  cat(" n.cond=", fit$n.cond, "\n")
}
cat("resid head (CSS):", paste(fmt(head(a$CSS$residuals, 3)), collapse = ", "), "\n")
cat("resid tail (CSS):", paste(fmt(tail(a$CSS$residuals, 2)), collapse = ", "), "\n")
cat("nresid length CSS:", length(a$CSS$residuals), "\n")

cat("\n=== arma22 (T=300) CSS / CSS-ML ===\n")
for (m in c("CSS", "CSS-ML")) {
  fit <- arima(arma22, order = c(2, 0, 2), method = m, include.mean = FALSE)
  cat(m, ": coef=", paste(fmt(fit$coef), collapse = ", "),
      " sigma2=", fmt(fit$sigma2),
      " loglik=", fmt(as.numeric(logLik(fit))),
      " n.cond=", fit$n.cond, "\n")
}

cat("\n=== arma21 (T=300, p=2 q=1) CSS — AR2 裁决: n.cond 实测 ===\n")
for (m in c("CSS", "CSS-ML")) {
  fit <- arima(arma21, order = c(2, 0, 1), method = m, include.mean = FALSE)
  cat(m, ": coef=", paste(fmt(fit$coef), collapse = ", "),
      " sigma2=", fmt(fit$sigma2),
      " loglik=", fmt(as.numeric(logLik(fit))),
      " n.cond=", fit$n.cond, "\n")
}

cat("\n=== arma12 (T=300, p=1 q=2) CSS — AR2 定案: max(p,q)=2 vs p=1 ===\n")
for (m in c("CSS", "CSS-ML")) {
  fit <- arima(arma12, order = c(1, 0, 2), method = m, include.mean = FALSE)
  cat(m, ": coef=", paste(fmt(fit$coef), collapse = ", "),
      " sigma2=", fmt(fit$sigma2),
      " loglik=", fmt(as.numeric(logLik(fit))),
      " n.cond=", fit$n.cond, "\n")
}

cat("\n=== arima111d: level 数据, d=1 ===\n")
# 语义裁决 (2026-08-18 实测): stats::arima 对 d≥1 默认 include.mean=FALSE
#   → 漂移未建模 → AR 多项式近单位根吸收漂移 (伪 φ=0.998, 退化路径;
#   对照 C++ arima_fit d=1 include.mean=false 行为)
# include.mean=TRUE → 差分序列均值吸收漂移 (与 statsmodels demean 对齐)
for (m in c("CSS", "CSS-ML")) {
  fit <- arima(arima111d_level, order = c(1, 1, 1), method = m)
  cat(m, " mean=FALSE: coef=", paste(fmt(fit$coef), collapse = ", "),
      " sigma2=", fmt(fit$sigma2),
      " loglik=", fmt(as.numeric(logLik(fit))),
      " n.cond=", fit$n.cond, "\n")
}
for (m in c("CSS", "CSS-ML")) {
  fit <- arima(arima111d_level, order = c(1, 1, 1), method = m,
               include.mean = TRUE)
  cat(m, " mean=TRUE: coef=", paste(fmt(fit$coef), collapse = ", "),
      " sigma2=", fmt(fit$sigma2),
      " loglik=", fmt(as.numeric(logLik(fit))),
      " n.cond=", fit$n.cond, "\n")
}
# forecast::Arima with drift (AR5): CSS-ML 语义
library(forecast)
fa <- Arima(arima111d_level, order = c(1, 1, 1), include.drift = TRUE,
            method = "CSS-ML")
cat("forecast::Arima drift: coef=", paste(fmt(fa$coef), collapse = ", "),
    " sigma2=", fmt(fa$sigma2),
    " loglik=", fmt(as.numeric(logLik(fa))),
    " drift=", fmt(fa$coef["drift"]), "\n")
fa_css <- Arima(arima111d_level, order = c(1, 1, 1), include.drift = TRUE,
                method = "CSS")
cat("forecast::Arima drift CSS: coef=", paste(fmt(fa_css$coef), collapse = ", "),
    " sigma2=", fmt(fa_css$sigma2), "\n")

File: tests/fixtures/timeseries/verify_arima.py
# -*- coding: utf-8 -*-
# verify_arima.py - ARIMA 基准生成: statsmodels innovations_mle (Phase 7C M1)
#
# 对照点 (spec §1.3):
#   ARIMA innovations vs statsmodels innovations_mle (无缺失非季节) 1e-10
#
# 语义对齐 (estimators/innovations.py 一手, 2026-08-18):
#   - statsmodels: 差分 d → demean=True → HR 起始 → scipy minimize
#     全参数 (φ,θ,σ²) on −arma_loglike
#   - C++: 差分 d → demean → {HR,0,扰动} SLSQP on 集中化 nll (σ² 解析)
#   - 同一似然面; 落点差 = 优化器容差层 (实测落档, 7B GM 先例口径)
#   ⚠️ d=0 夹具数据非零均值 (测 demean 语义); C++ innovations_mle(demean=True)
#
# 用法: python verify_arima.py   (生成 arima_smoke_data.csv + 基准 JSON)
import json
import os

import numpy as np

HERE = os.path.dirname(os.path.abspath(__file__))


def gen_arma11(T=300, seed=42):
    """ARMA(1,1): φ=0.5, θ=−0.4, σ=1, burn-in 100 (RandomState 跨语言固定)"""
    rng = np.random.RandomState(seed)
    phi, theta, e = 0.5, -0.4, rng.standard_normal(T + 100)
    z = np.zeros(T + 100)
    for t in range(1, T + 100):
        z[t] = phi * z[t - 1] + e[t] + theta * e[t - 1]
    return z[100:]


def gen_arma22(T=300, seed=43):
    """ARMA(2,2): φ=(0.3,−0.2), θ=(−0.5,0.3) — 多峰逃逸夹具 (AR7)"""
    rng = np.random.RandomState(seed)
    phi, theta = [0.3, -0.2], [-0.5, 0.3]
    e = rng.standard_normal(T + 100)
    z = np.zeros(T + 100)
    for t in range(2, T + 100):
        z[t] = (phi[0] * z[t - 1] + phi[1] * z[t - 2] + e[t]
                + theta[0] * e[t - 1] + theta[1] * e[t - 2])
    return z[100:]


def gen_arma21(T=300, seed=45):
    """ARMA(2,1): p≠q — AR2 裁决夹具 (R CSS n.cond = d+max(p,q) vs d+p)"""
    rng = np.random.RandomState(seed)
    phi, theta = [0.4, -0.25], [-0.6]
    e = rng.standard_normal(T + 100)
    z = np.zeros(T + 100)
    for t in range(2, T + 100):
        z[t] = (phi[0] * z[t - 1] + phi[1] * z[t - 2] + e[t]
                + theta[0] * e[t - 1])
    return z[100:]


def gen_arma12(T=300, seed=46):
    """ARMA(1,2): p<q — AR2 定案夹具 (max(p,q)=2 vs p=1 → n.cond 1 or 2)"""
    rng = np.random.RandomState(seed)
    phi, theta = [0.35], [-0.5, 0.25]
    e = rng.standard_normal(T + 100)
    z = np.zeros(T + 100)
    for t in range(2, T + 100):
        z[t] = (phi[0] * z[t - 1] + e[t]
                + theta[0] * e[t - 1] + theta[1] * e[t - 2])
    return z[100:]


def gen_arima111_drift(T=300, seed=44):
    """ARIMA(1,1,1)+drift: Δy = 0.3 + ARMA(1,1)"""
    rng = np.random.RandomState(seed)
    d = gen_arma11(T, seed) + 0.3
    return np.cumsum(d)


def main():
    from statsmodels.tsa.arima.estimators.innovations import innovations_mle

    fixtures = {
        "arma11": {"data": gen_arma11(), "order": (1, 0, 1), "demean": True},
        "arma22": {"data": gen_arma22(), "order": (2, 0, 2), "demean": True},
        "arma21": {"data": gen_arma21(), "order": (2, 0, 1), "demean": True},
        "arma12": {"data": gen_arma12(), "order": (1, 0, 2), "demean": True},
        # innovations+drift: statsmodels 无 drift 参数 → demean=True 吸收
        # (drift ≈ 差分序列均值; C++ 侧 innovations_mle(demean=True) 对齐)
        "arima111d": {"data": np.diff(gen_arima111_drift()),
                      "order": (1, 0, 1), "demean": True},
    }

    out = {}
    for name, fx in fixtures.items():
        p, other = innovations_mle(fx["data"], order=fx["order"],
                                   demean=fx["demean"])
        llf = None
        try:
            from statsmodels.tsa.innovations.arma_innovations import (
                arma_loglike,
            )
            z = fx["data"] - fx["data"].mean()
            llf = float(arma_loglike(
                z, ar_params=-np.asarray(p.reduced_ar_poly.coef[1:]),
                ma_params=np.asarray(p.reduced_ma_poly.coef[1:]),
                sigma2=float(p.sigma2)))
        except Exception as exc:  # noqa: BLE001
            llf = None
            print(f"{name}: llf eval failed: {exc}")
        out[name] = {
            "phi": [float(v) for v in p.ar_params],
            "theta": [float(v) for v in p.ma_params],
            "sigma2": float(p.sigma2),
            "loglik": llf,
            "demean": fx["demean"],
            "order": list(fx["order"]),
            "nobs": int(len(fx["data"])),
            "converged": bool(other["minimize_results"].success),
        }
        print(name, json.dumps(out[name], indent=1))

    # 数据 CSV (%.17g 全精度, C++/R 同源消费; 各列同长 T=300;
    # arima111d 存 level, 差分由消费方自行 diff — 列长一致原则)
    with open(os.path.join(HERE, "arima_smoke_data.csv"), "w") as f:
        cols = {"arma11": fixtures["arma11"]["data"],
                "arma22": fixtures["arma22"]["data"],
                "arma21": fixtures["arma21"]["data"],
                "arma12": fixtures["arma12"]["data"],
                "arima111d_level": gen_arima111_drift()}
        keys = list(cols)
        lens = {len(v) for v in cols.values()}
        assert len(lens) == 1, f"column length mismatch: {lens}"
        n = lens.pop()
        for i in range(n):
            f.write(",".join("%.17g" % cols[k][i] for k in keys) + "\n")
        f.write("#columns=" + ",".join(keys) + "\n")

    with open(os.path.join(HERE, "arima_statsmodels_baselines.json"), "w") as f:
        json.dump(out, f, indent=1)
    print("written: arima_statsmodels_baselines.json")


if __name__ == "__main__":
    main()
